### DISCUSSION TITLE
TabBar: dotted border around clicked tab

### DISCUSSION THREAD

**Discussion by martinkrulltott** [dhis2/notes#223] [Q&A] [answered]
Since ui@6.1.7, when clicking a ´Tab´ in a ´TabBar´, a dotted border is shown around the label. As seen in the screenshot and example sandbox below:

https://codesandbox.io/s/ui-tabs-q3nye?file=/src/App.js

![image](https://user-images.githubusercontent.com/12590483/107617559-6d305680-6c50-11eb-83fc-ba9a3b222ff2.png)

**Comment by amcgee** [comment, marked as answer]
@martinkrulltott this is actually by design - see [this issue](https://jira.dhis2.org/browse/LIBS-77) and [this pr](https://github.com/dhis2/ui/pull/482), and you can see this in the [release notes of the v6.1.7 release](https://github.com/dhis2/ui/releases/tag/v6.1.7)

This is an a11y (accessibility) feature of browsers which allows visual feedback when tabbing through focusable components.

**Reply by martinkrulltott** [reply]
I understand the accessibility part of the feature for tabbing, I just didn't expect it on click, as I haven't seen that feature anywhere else in neither Data Visualizer nor ui. But as this is per design then this question is ✅

**Comment by amcgee** [comment]
It does seem like there could be a valid reason to unfocus the tab after it has been *clicked* (since that didn't depend on pre-existing focus).  There appears to be some debate about how or if to do this, but blurring on `mouseup` seems like a nice compromise to me (one of many solutions from [StackOverflow](https://stackoverflow.com/a/30119360))

[Here's an example](https://codesandbox.io/s/ui-tabs-forked-c27i7) using a hack to add the event to the buttons.

@cooper-joe @mediremi @paschalidi any thoughts on changing this behavior from a UI or accessibility standpoint?

**Reply by amcgee** [reply]
One issue here is that anything else which previously had focus will lose focus, so tabstops will reset to the beginning

**Reply by amcgee** [reply]
We could of course also do this on `mousedown` to avoid the flash of outline... but that might be worse?  Using `mouseup` if you depress your mouse on the button, then move it somewhere else before releasing, the half-clicked button remains focused

**Reply by amcgee** [reply]
@cooper-joe another thought here - we actually already have a "hover" style for tabs which I think might be appropriate as the "focus" style as well - when you tab through it could behave similarly to when hovering over the tabs.  This could allow us to have a "subtle" active+focused style (blue bottom border and gray background color) when you click a tab, allowing us to keep it focused and for the user to continue tabbing from there.

**Reply by mediremi** [reply]
I like the idea of having a focus style similar to the hover style. We might want to think about disabling transitions for the focus styles though as I find they can make keyboard navigation feel sluggish.

**Reply by paschalidi** [reply]
Here is what is suggested from an a11y point of view.
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role
Long story short there is some sort of outline when you click the tab. Also optionally you can navigate using the <- and -> arrow to the previous and next tab. They say that ideally when you press tab your focus moves inside the panel of the tab.

![image](https://user-images.githubusercontent.com/4181674/107631548-14b38600-6c5d-11eb-9f82-150bfee98918.png)


----

Another example can be seen in Chakra who they call themselves an "accessible component library"
https://chakra-ui.com/docs/disclosure/tabs#usage

**Reply by cooper-joe** [reply]
I agree that using a similar style for `hover` and `focus` is a useful way of making this seamless. I think this would be a good starting point for the aesthetic part of this discussion.

**Comment by amcgee** [comment]
If you try interacting with GitHub interface (see the `Write` and `Preview` tabs when writing a comment) is pretty enlightening - they don't have separate tabstops for each tab but rather a single one for the **tab bar** and support arrow keys to navigate between tabs.  Also, when you click on one of the tabs it moves focus to whatever the first focusable element is **within that tab** which I think makes sense.

Navigating through the "tab-like" sidebar options on the lefthand side of the Codesandbox linked above - if you click on one of the tabs, it is selected.  The Active and Focused styles are identical, so this is pretty seamless.  If you tab after clicking on one of the tab buttons, the focus starts from whichever tab you clicked which I think is quite nice (and is broken by the `blur` solution mentioned above)

**Reply by amcgee** [reply]
Notably - neither of these examples reset focus to the top of the page after clicking a tab/button

**Reply by cooper-joe** [reply]
Reading into this a bit more, I think the tabbing from the active `tab` in `tab list` to the `tab panel` (or, `content`) is the [recommended design pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel). Arrow keys are used to navigate between `tabs` in the `tab list`. 
I think this would be the interaction pattern to support if we want to change the current behaviour.

**Reply by varl** [reply]
We need to pull this out into a JIRA issue and link it to an Epic for adding WAI-ARIA to components as we should not do the work in relation to this question as a quick fix.

**Reply by varl** [reply]
I've created an issue for the TabBar a11y here: https://jira.dhis2.org/browse/LIBS-124

It is part of an Epic that covers accessibility for all existing components. Future components should consider a11y from the start.

**Comment by janhenrikoverland** [comment]
@cooper-joe @varl @amcgee Does anyone have any experience with e.g. using `:focus-visible` ? It could make sense to show such focus indicators only for keyboard interaction. That seems to be the intention as well: `the W3C proposal mentions that :focus-visible should be preferred over :focus except on elements that expect a keyboard input (e.g. text field, contenteditable)`.

Btw, at the moment we have four different focus indicators... dotted, blue, green and yellow border.

![image (8)](https://user-images.githubusercontent.com/1010094/107924292-04f4b400-6f73-11eb-85e3-c1dff42b603a.png)

---

![image (7)](https://user-images.githubusercontent.com/1010094/107924296-058d4a80-6f73-11eb-9e80-1116dd30fda4.png)

---

![image (6)](https://user-images.githubusercontent.com/1010094/107924297-058d4a80-6f73-11eb-9d91-1aea2b0c19d0.png)

---

![image (5)](https://user-images.githubusercontent.com/1010094/107924556-6ddc2c00-6f73-11eb-9d0a-82bacca4c192.png)
